### PR TITLE
Add a OSX-specific and Mono FileSystemWatcher wrapper

### PR DESCRIPTION
**Pull request by luhenry**
This integrates the FSW from CoreFX which is OSX specific.

### CODE REVIEW

File: main/build/MacOSX/System_Native.cpp [hunk 1]
```diff
@@ -0,0 +1,636 @@
+
```
Comment (Therzok) [inline main/build/MacOSX/System_Native.cpp]: This file is a great idea!

File: main/src/core/MonoDevelop.Core/MonoDevelop.FSW/FileSystemWatcher.cs [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+
+		enum Platform
+		{
+			OSX,
+			Mono,
+		}
+
+		static FileSystemWatcher ()
+		{
+			// FIXME add detection for Mono or OSX
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.FSW/FileSystemWatcher.cs]: For self reference: Maybe the whole thing can go into its separate repo, and try a `EntryPointNotFoundException` check on whether one of the System.Native pinvokes fails.

**Comment by Therzok** [comment]
All in all, this looks amazing!

File: main/src/core/MonoDevelop.Core/MonoDevelop.FSW/OSX/FileSystemWatcher.OSX.cs [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.FSW/OSX/FileSystemWatcher.OSX.cs]: The above header should be removed from all files we are taking directly from .NET Core.

**Comment by ludovic-henry** [comment]
@slluis I updated according to your review.

@Therzok I added the `HasOSXSupport` P/Invoke which will test if the current native executable has the `SystemNative_*` functions (https://github.com/mono/monodevelop/pull/1906/files#diff-5a37ecc510b9e231ec56752bd1e86b3dR48)

**Comment by luhenry** [comment]
@Therzok anything blocking this from getting merged?

**Comment by slluis** [comment]
Other than the public class issue it looks ok to me. Now we have to find a good way of using it.

**Comment by luhenry** [comment]
@slluis I fixed the public class issue.

**Comment by Therzok** [comment]
@slluis my plan of attack would be to split this code into its own repo, so we can submodule it where necessary and use it in the repos we need. How does it sound?

**Comment by luhenry** [comment]
@Therzok @slluis  IMO it's definitely overkill to split it in its own repo, as the long term plan is to have this FSW from CoreFX directly in Mono. Having copied the source inside XS is itself a gross hack, that shouldn't be there in the first place, but which is the fastest way to get where we want to get.

My advice would simply be to copy/paste it where you need it, keep the wrapper around the OSX and Mono "backends", and once the CoreFX FSW is in Mono, simply remove the copied OSX version, and remove/inline the wrapper and call the Mono implementation directly.

**Pull request closed without merging**